### ISSUE TITLE
art_integration_in_server_test is broken in oscillatord version using serial interface

### ISSUE THREAD

**Issue by leoleovich** [Orolia2s/oscillatord#76]
It seems like `art_integration_in_server_test` version using serial interface to mRO50 is currently broken. Here is the observation summary:

Items | Timecard Driver | ART Firmware | Oscillatord | disciplining-minipod | art_integration_in_server_test
-- | -- | -- | -- | -- | --
1 | 3rd Aug | 0.0.15 | 3.3.6 | 3.3.6 | FAILED
2 | 3rd Aug | 0.0.13 | 3.3.6 | 3.3.6 | FAILED
3 | 3rd Aug | 0.0.15 | 3.2.4 | 3.2.4 | PASSED
4 | 3rd Aug | 0.0.13 | 3.2.4 | 3.2.4 | PASSED
5 | 3rd Aug | 0.0.13 | 3.3.7 | 3.3.7 | FAILED
6 | 3rd Aug | 0.0.15 | 3.3.7 | 3.3.7 | FAILED

**Comment by leoleovich** [comment]
cc @CharlesParent :)

**Comment by KritsadaM** [comment]
cc @jachenilleau 
As I reported you on email
Here is failure log on art_integration_in_server_test v3.3.6
https://gist.github.com/KritsadaM/b5c59d16c9b55e4baf30db81669e06b2

**Comment by CharlesParent** [comment]
Hi @leoleovich !

Does art_integration_in_server_test works if you run oscillatord for a few minutes before ?

If so that means that serial device is not activated by art_integration_in_server_test (it needs to have the ioctl device to do so), just like in src/oscillators/mro50.c:

fd = open(devices_path->mro_path, O_RDWR);
	if (fd < 0) {
		log_error("Could not open mRo50 device\n");
		goto error;
	}
	mRo50->osc_fd = fd;

	/* Activate serial in order to use mro50-serial device */
	uint32_t serial_activate = 1;
	ret = ioctl(fd, MRO50_BOARD_CONFIG_WRITE, &serial_activate);
	if (ret != 0) {
		log_error("Could not activate mro50 serial");
		goto error;
	}

You may add this in test software though I would advise to just get rid of this last IOCTL. Hervé should release a new firmware to activate serial communication by default and then all ioctls could be erased from both ART software and driver

**Comment by KritsadaM** [comment]
@CharlesParent 

After update the code per your suggestion.
"art_integration_in_server_test" can get pass test but it failed during "Started oscillatord@ocp0 service"

https://gist.github.com/KritsadaM/60df348f7a230223a427c837f16ccd14

Could you please help to review ?

**Comment by CharlesParent** [comment]
Could you retry and give me the jounarl logs or oscillatord during test ?

**Comment by KritsadaM** [comment]
Hi @CharlesParent,

There is no oscillatord log during art_integration_in_server_test
As I tried to troubleshoot the problem. it seems like there is some failure on "Initiate a connection to the server" code portion

However, oscillatord service can run without problem.
Below is 5 mins test time for your reference.

[oscillatord.log](https://github.com/Orolia2s/oscillatord/files/9402431/oscillatord.log)

```
static struct json_object *send_monitoring_request(int socket_port, enum monitoring_request request) {
    int sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if (sockfd == -1)
    {
        log_error("Could not connect to socket !");
        log_error("FAIL");
        return NULL;
    }

    struct sockaddr_in server_addr;
    memset(&server_addr, 0, sizeof(server_addr));
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(socket_port);
    server_addr.sin_addr.s_addr = inet_addr("0.0.0.0");

    /* Initiate a connection to the server */
    int ret = connect(sockfd, (struct sockaddr *)&server_addr, sizeof(server_addr));
==================================================
# ret will get '-1' and fail at this point 
==================================================
    if (ret == -1)
    {
        log_error("Could not connect to socket !");
        log_error("FAIL");
        return NULL;
    }
    struct json_object *obj = json_send_and_receive(sockfd, request);
    log_info(json_object_to_json_string(obj));
    close(sockfd);
    return obj;

}
```

**Comment by CharlesParent** [comment]
@KritsadaM it's strange because oscillatord should be started by the test procedure (in oscillatord_activate_service), can you trave what happen ?

**Comment by KritsadaM** [comment]
Since there is no error message occur in oscillatord_activate_service, I assume oscillatord has been activated without problem.
I just add debug message which line cause software raise an error

```console
14:51:24 DEBUG OCP name is ocp0
14:51:24 DEBUG ocp number is 0
14:51:24 INFO  Starting Phase error limit test
14:51:24 INFO  Started oscillatord@ocp0 service
14:51:29 DEBUG ret : -1
14:51:29 DEBUG sockfd : 1
14:51:29 DEBUG server_addr : 0.0.0.0
14:51:29 ERROR Could not connect to socket !
14:51:29 ERROR FAIL
14:51:29 ERROR Could not get disciplining data
14:51:29 ERROR Card did not passed phase error limit test even after calibration
14:51:29 ERROR Test FAILED !
14:51:29 INFO  Stopped oscillatord@ocp0 service
```

```
    /* Initiate a connection to the server */
    int ret = connect(sockfd, (struct sockaddr *)&server_addr, sizeof(server_addr));
    char *ip = inet_ntoa(server_addr.sin_addr);
    log_debug("ret : %d",ret);
    log_debug("sockfd : %d",sockfd);
    log_debug("server_addr : %s",ip);
    if (ret == -1)
    {
        log_error("Could not connect to socket !");
        log_error("FAIL");
        return NULL;
    }
    struct json_object *obj = json_send_and_receive(sockfd, request);
    log_info(json_object_to_json_string(obj));
    close(sockfd);
    return obj;
```

https://gist.github.com/KritsadaM/efa34c8ea1fae109c30fdb6679279842

**Comment by CharlesParent** [comment]
Could you please log the socket port please ? 

If there are no mistake then I think there is an issue woth how oscillatord service isrance is started

**Comment by KritsadaM** [comment]
@CharlesParent 
Here is a result

``` console
01:20:24 DEBUG OCP name is ocp0
01:20:24 DEBUG ocp number is 0
01:20:24 INFO  Starting Phase error limit test
01:20:24 INFO  Started oscillatord@ocp0 service
01:20:29 DEBUG ret : -1
01:20:29 DEBUG sockfd : 1
01:20:29 DEBUG server_addr : 0.0.0.0
01:20:29 DEBUG socket_port : 2970
01:20:29 DEBUG Listening on ip 0.0.0.0 and port 2970

01:20:29 ERROR Could not connect to socket !
01:20:29 ERROR FAIL
01:20:29 ERROR Could not get disciplining data
01:20:29 ERROR Card did not passed phase error limit test even after calibration
01:20:29 ERROR Test FAILED !
01:20:29 INFO  Stopped oscillatord@ocp0 service
```

``` c
    /* Initiate a connection to the server */
    int ret = connect(sockfd, (struct sockaddr *)&server_addr, sizeof(server_addr));
    char *ip = inet_ntoa(server_addr.sin_addr);
    int port = ntohs(server_addr.sin_port);
    log_debug("ret : %d",ret);
    log_debug("sockfd : %d",sockfd);
    log_debug("server_addr : %s",ip);
    log_debug("socket_port : %d",port);
    log_debug("Listening on ip %s and port %d\n", ip, port);
    if (ret == -1)
    {
        log_error("Could not connect to socket !");
        log_error("FAIL");
        return NULL;
    }
    struct json_object *obj = json_send_and_receive(sockfd, request);
    log_info(json_object_to_json_string(obj));
    close(sockfd);
    return obj;
```

Any comments, please feel let me know.

**Comment by CharlesParent** [comment]
Ok so port looks good so I think there might be an issue with the function starting / stopping oscillatord service.

This needs further investigation, in the meantime you could start oscillatord manually and check that phase error is below 100ns (that is what the test is doing actually)

**Comment by KritsadaM** [comment]
[oscillatord.log](https://github.com/Orolia2s/oscillatord/files/9429579/oscillatord.log)

The results of 10 mins testing of oscillatord service is 0~+/-5 ns
Could you please review the result ?

Any comments, please feel free to let me know.

**Comment by CharlesParent** [comment]
From what I read it looks good !

**Comment by jachenilleau** [comment]
Hi, I've fixed a couple of issues. 
All changes are in branch : https://github.com/Orolia2s/oscillatord/tree/art_integration_in_server_fix
Can you check and report from your side @KritsadaM ?
Thanks

**Comment by KritsadaM** [comment]
@jachenilleau 

After checkout your branch and build, we can get pass result from art_integration_in_server_test
[art_integration_in_server_test.log](https://github.com/Orolia2s/oscillatord/files/9474470/art_integration_in_server_test.log)

Thank you for your support

**Comment by jachenilleau** [comment]
Fixed in v3.3.8
